lerna-util-sequence/src/main/scala/lerna/util/sequence/SequenceStore.scala

```diff
@@ -40,16 +41,17 @@ private[sequence] object SequenceStore extends AppTypedActorLogging {
 
     // SupervisorStrategy
     // see: https://docs.datastax.com/en/developer/java-driver/3.6/manual/retries/#retry-policy
-    behavior = Behaviors.supervise(behavior).onFailure[NoHostAvailableException](SupervisorStrategy.restart)
-    behavior = Behaviors.supervise(behavior).onFailure[UnsupportedFeatureException](SupervisorStrategy.restart)
+    // TODO 対応する例外に更新する
+    // behavior = Behaviors.supervise(behavior).onFailure[NoHostAvailableException](SupervisorStrategy.restart)
+    // behavior = Behaviors.supervise(behavior).onFailure[UnsupportedFeatureException](SupervisorStrategy.restart)
     // 一時的にレプリカが処理できなくなっているだけなので、Cassandra サイドで回復することを期待する
     behavior = Behaviors.supervise(behavior).onFailure[ReadTimeoutException](SupervisorStrategy.resume)
     // 一時的にレプリカが処理できなくなっているだけなので、Cassandra サイドで回復することを期待する
     behavior = Behaviors.supervise(behavior).onFailure[WriteTimeoutException](SupervisorStrategy.resume)
     // コーディネーターに何らかの問題が起きている可能性がある。再接続して回復することを期待する
-    behavior = Behaviors.supervise(behavior).onFailure[OperationTimedOutException](SupervisorStrategy.restart)
+    // behavior = Behaviors.supervise(behavior).onFailure[OperationTimedOutException](SupervisorStrategy.restart)
     // コネクションに問題がある。再接続して回復することを期待する
-    behavior = Behaviors.supervise(behavior).onFailure[ConnectionException](SupervisorStrategy.restart)
+    // behavior = Behaviors.supervise(behavior).onFailure[ConnectionException](SupervisorStrategy.restart)
     behavior = Behaviors.supervise(behavior).onFailure[Exception](SupervisorStrategy.restart)
```

Comment:
@xirc 
ドライバーがスローする例外の体系が変わっているなら、追加で `resume` にできそうな例外が無いか確認しておきたいですね。
確認をお願いしても良いですか？

`SequenceStore` の現在の実装でも restart すると Cassandra の Session が再作成されるので、従来どおり、接続しなおしたほうが回復が期待できる例外に関しては  restart で、接続しなおす必要が無さそうな例外に関しては resume という判断基準で良いと思います。

> Exception をまとめて restart しているため、 resume している supervise だけを残してもよさそうに思います。

これに関しては、 restart のほうが良い（resume だと問題がありそう）と判断した経緯があることを示すために検討した結果 restart にした例外に関しては `onFailure` を明示的に書いておいたほうが良さそうな気がしています。resume と restart がどういう意図で使い分けられているのかが読み取りやすそうに思うので。

もしくは、前述した判断基準をコメントとして書いておくという方法でも良いかもしれません。

> 接続しなおしたほうが回復が期待できる例外に関しては  restart で、接続しなおす必要が無さそうな例外に関しては resume という判断基準で良いと思います。

Reply:
ありがとうございます。
判断基準をコメントに記載しました。
https://github.com/lerna-stack/lerna-app-library/pull/54/commits/906402373e020780691e54ed6b5d6ed189536bfe

Reply:
@negokaz 

`resume` を増やすかどうかは、改善である認識でした。
この PR での改善はプランにありませんでしたが、対応することも良いかもしれません。

`ReadTimeoutException` の際に、 `resume` することで問題が発生しないのか、私には分かっていません。
`WriteTimeoutException` も同様です。

https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/retries/#built-in-retry-policies
https://docs.datastax.com/en/developer/java-driver/3.11/manual/retries/#on-read-timeout
> This could be due to a temporary overloading of these replicas, or even that they just failed or were turned off. 

新旧ドライバのドキュメントには、`ReadTimeoutException` が発生するのは一時的な過負荷の可能性があると記載されています。
なぜ一時的な過負荷になっているのかを発生した例外から判断することは困難だと思いますが、
障害の連鎖を防ぐ観点から考えると、backoff をつけて `restart` する方が良いようにも思います。

また、Actor レベルで `resume`, `restart` のどちらを選ぶかに関わらず、
`DefaultRetryPolicy` を使っていれば冪等であるリクエストは最大1回までリトライされます。
ほとんどの一時的なネットワーク障害は、このリトライで十分に対処できそうにも思います。
https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/retries/#built-in-retry-policies

`ReadTimeoutException` と `WriteTimeoutException` だけ特別に`resume` を採用している理由は、"コネクションを再生成したくない"という理由のみでしょうか？それ以外の理由もありますか？

Reply:
@xirc 

> 障害の連鎖を防ぐ観点から考えると、backoff をつけて restart する方が良いようにも思います。

なるほど、確かにその観点だと backoff で間隔をあけてあげたほうが良さそうですね。
restart すると、`SequenceStore` に stash したコマンドやコネクションの再確立が行われるのが少し気になりますが。。

> ReadTimeoutException と WriteTimeoutException だけ特別にresume を採用している理由は、"コネクションを再生成したくない"という理由のみでしょうか？それ以外の理由もありますか？

実装した当時の理由はそれだけでした。コネクションの再確立が時間的にもリソース的にもコストがかかるというだろうという予想のもと、コネクションに問題がないのならそれを破棄せずに使いまわしたほうが採番の可用性が高くなるだろうという思いからです。

追加で検討に入れておきたいのが、今の実装だと `SequenceStore` が採番予約を処理中は `SequenceFactoryWorker` 側で追加の採番予約をしない実装にしていましたが、 #49 の改善の中で `SequenceStore` が処理中かどうかを  `SequenceFactoryWorker`  で管理するのはやめ、とりあえず採番予約のメッセージを送りつけるように変更しようかと考えていました。その場合、`SequenceStore` 側で採番予約のコマンドがクエリの実行中に stash されることになるので、restart によってその stash されたコマンドが消えてしまっても問題ないかは確認が必要そうです。

Reply:
@negokaz 

1. コネクションの再確立は、 https://github.com/lerna-stack/lerna-app-library/pull/54#discussion_r700720631 のトピックで対応する予定の 「`datastax-java-driver` の代わりに `akka-stream-alpakka-cassandra` に依存する」ことによって、改善されるかもしれません。
2. `stash` してもよいかどうかは、そのほかの restart も関連していると思うので、  #49 の対応で検討するほうが良さそうに思いました。

https://github.com/lerna-stack/lerna-app-library/pull/54/files#r697175861 にて、少なくとも `resume` する既存の例外の意味が変わっていないことは確認してあります。 `resume` するかどうか、`resume` する例外を増やすかどうかは、より発展的な改善に相当するように思います。私の考えでは、ドライババージョンアップに起因していないと思われるこのような発展的な改善は、この PR に含めない方が良さそうに思いますが、何か見落としている観点はありますか？

Reply:
@xirc 

> stash してもよいかどうかは、そのほかの restart も関連していると思うので、 #49 の対応で検討するほうが良さそうに思いました。

そうですね。この PR では `SequenceStore` についてだけ検討しましょう。 

> resume するかどうか、resume する例外を増やすかどうかは、より発展的な改善に相当するように思います。私の考えでは、ドライババージョンアップに起因していないと思われるこのような発展的な改善は、この PR に含めない方が良さそうに思いますが、何か見落としている観点はありますか？

改善というよりかは、現状維持を目指しています。
気になっているポイントは `SequenceStore` が `restart` するケースがバージョンアップによって余計に増えていないかという点です。
ドライババージョンアップによって新しく追加された、`resume` によって対処可能な例外のハンドリングを省略したことによって `SequenceStore` の可用性が従来よりも下がってしまう可能性があるのなら、バージョンアップの時点で対処しておきたいなと思っています。
従来から存在し、`restart` で対処されている例外に関してはそのままで良いと思っています。

Reply:
@negokaz 

次のコミットで対応しました。
> Handle exceptions explicitly 785eee88aca7d688198dcbd9ccc410cd0c4680a7

例外を、新旧で 1対1 対応させるのは難しかったため、主に新ドライバの例外について確認しました。
https://docs.datastax.com/en/developer/java-driver/4.6/manual/core/retries/#retries
https://docs.datastax.com/en/developer/java-driver/3.11/manual/retries/

次の 3つが `resume` として 追加されています。
```scala
    behavior = Behaviors.supervise(behavior).onFailure[UnavailableException](SupervisorStrategy.resume)
    behavior = Behaviors.supervise(behavior).onFailure[ReadFailureException](SupervisorStrategy.resume)
    behavior = Behaviors.supervise(behavior).onFailure[WriteFailureException](SupervisorStrategy.resume)
```


対応付けられそうなものは次の通りです。

## `UnavailableException`
* [UnavailableException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/UnavailableException.html)
* [UnavailableException (DataStax Java driver for Apache Cassandra(R) 4.6.0 API)](https://docs.datastax.com/en/drivers/java/4.6/com/datastax/oss/driver/api/core/servererrors/UnavailableException.html)

https://docs.datastax.com/en/developer/java-driver/4.6/manual/core/retries/#on-unavailable
https://docs.datastax.com/en/developer/java-driver/3.11/manual/retries/#on-unavailable

> If the policy rethrows the error, the user code will get an UnavailableException. You can inspect the exception’s fields to get the amount of replicas that were known to be alive when the error was triggered, as well as the amount of replicas that where required by the requested consistency level.


変更前のコードでは、例外が処理されていません。
変更後のコードでは、resume しています。
説明を読む限り、Coordinator とやり取りできているため、Cassandra に再接続する必要はないように思います。


## `NoHostAvailableException` => `AllNodesFailedException`

ほとんど同じ意味だと思われます。
引き続き restart しますが、おそらくランダムバックオフをつけるほうが良いかもしれません。

## `UnsupportedFeatureException` => ???
* [UnsupportedFeatureException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/UnsupportedFeatureException.html)

対応を見つけられませんでしたが、サポートされていない機能を使用しようとしているため、
 resume で復旧できる可能性は低いように思います。

## `OperationTimedOutException` => `DriverTimeoutException`
* [OperationTimedOutException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/OperationTimedOutException.html)
* [DriverTimeoutException (DataStax Java driver for Apache Cassandra(R) 4.6.0 API)](https://docs.datastax.com/en/drivers/java/4.6/com/datastax/oss/driver/api/core/DriverTimeoutException.html)

ほとんど同じ意味だと思われます。
Coordinator が時間内に応答していません。従来通り restart します。

## `ConnectionException` => ` ClosedConnectionException`
* [ConnectionException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/ConnectionException.html)
* [ClosedConnectionException (DataStax Java driver for Apache Cassandra(R) 4.6.0 API)](https://docs.datastax.com/en/drivers/java/4.6/com/datastax/oss/driver/api/core/connection/ClosedConnectionException.html)

ほとんど同じ意味だと思われます。
コネクションが閉じられているため、 引き続き restart します。

## その他
一部処理は、RetryPolicy に依存します。
新旧 DefaultRetryPolicy は次の URL から確認できます。
* https://github.com/datastax/java-driver/blob/3.7.1/driver-core/src/main/java/com/datastax/driver/core/policies/DefaultRetryPolicy.java
* https://github.com/datastax/java-driver/blob/4.6.0/core/src/main/java/com/datastax/oss/driver/internal/core/retry/DefaultRetryPolicy.java

Reply:
## `ReadFailureException`
[ReadFailureException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/ReadFailureException.html)
[ReadFailureException (DataStax Java driver for Apache Cassandra(R) 4.6.0 API)](https://docs.datastax.com/en/drivers/java/4.6/com/datastax/oss/driver/api/core/servererrors/ReadFailureException.html)

> A non-timeout error during a read query.
> This happens when some of the replicas that were contacted by the coordinator replied with an error.

同じ意味だと思われます。
Cassandra に再接続する必要はないという観点では、resume できそうに思います。

## `WriteFailureException`
[WriteFailureException (DataStax Java Driver for Apache Cassandra - Binary distribution 3.11.0 API)](https://docs.datastax.com/en/drivers/java/3.11/com/datastax/driver/core/exceptions/WriteFailureException.html)
[WriteFailureException (DataStax Java driver for Apache Cassandra(R) 4.6.0 API)](https://docs.datastax.com/en/drivers/java/4.6/com/datastax/oss/driver/api/core/servererrors/WriteFailureException.html)

> A non-timeout error during a write query.
> This happens when some of the replicas that were contacted by the coordinator replied with an error.

同じ意味だと思われます。
Cassandra に再接続する必要はないという観点では、resume できそうに思います。

Reply:
`ReadFailureException` と `WriteFailureException` は Cassandra に再接続する必要ななさそうですが、タイムアウトエラーではないため、resume  するとCassandra クラスタを過負荷にする可能性があるかもしれません。個人的には `restart` の方がより良いように思いますが、基本方針である `Cassandra に再接続しなくても回復できそうな例外は resume する` を優先しました。

Reply:
@xirc 対応ありがとうございます！対処内容に問題無さそうです